### PR TITLE
Add Datastore Data project API

### PR DESCRIPTION
**Pull request by owainhunt**
Implemented the 'main' datastore API, ie everything on this page: https://cloud.google.com/datastore/docs/reference/data/rest/v1/projects

### CODE REVIEW

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    
+    public init(distinctOn: [PropertyReference]? = nil,
+        endCursor: String? = nil,
+        filter: Filter.TypedFilter? = nil,
+        kind: [KindExpression]? = nil,
+        limit: Int? = nil,
+        offset: Int? = nil,
+        order: [PropertyOrder]? = nil,
+        projection: [Projection]? = nil,
+        startCursor: String? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Just some spacing.

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    
+    init(distinctOn: [PropertyReference]? = nil,
+                 endCursor: String? = nil,
+                 filter: Filter? = nil,
+                 kind: [KindExpression]? = nil,
+                 limit: Int? = nil,
+                 offset: Int? = nil,
+                 order: [PropertyOrder]? = nil,
+                 projection: [Projection]? = nil,
+                 startCursor: String? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Spacing/indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        switch typedFilter {
+        case .composite(let filter):
+            self.init(compositeFilter: filter, propertyFilter: nil); return
+        case .property(let filter):
+            self.init(compositeFilter: nil, propertyFilter: filter); return
+        }
+    }
+    
+    init(compositeFilter: CompositeFilter? = nil,
+                 propertyFilter: PropertyFilter? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    /// The property to order by.
+    public let property: PropertyReference
+    
+    public enum Direction: String, RawRepresentable, GoogleCloudModel {
+        case ascending = "ASCENDING"
+        case descending = "DESCENDING"
+    }
+    
+    public init(direction: Direction = .ascending,
+                 property: PropertyReference) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        case lessThanOrEqual = "LESS_THAN_OR_EQUAL"
+        case greaterThan = "GREATER_THAN"
+        case greaterThanOrEqual = "GREATER_THAN_OR_EQUAL"
+        case equal = "EQUAL"
+        case hasAncestor = "HAS_ANCESTOR"
+    }
+    
+    public init(op: Operator? = nil,
+                 property: PropertyReference? = nil,
+                 value: Value? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        case noMoreReults = "NO_MORE_RESULTS"
+    }
+    
+    public init(endCursor: String? = nil,
+                 entityResultType: ResultType? = nil,
+                 entityResults: [EntityResult]? = nil,
+                 moreResults: MoreResultsType? = nil,
+                 skippedCursor: String? = nil,
+                 skippedResults: Int? = nil,
+                 snapshotVersion: String? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: Indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    /// Numbered binding site @1 references the first numbered parameter, effectively using 1-based indexing, rather than the usual 0.
+    /// For each binding site numbered i in queryString, there must be an i-th numbered parameter. The inverse must also be true.
+    public let positionalBindings: [GqlQueryParameter]?
+    /// A string of the format described [here](https://cloud.google.com/datastore/docs/reference/gql_reference).
+    public let queryString: String?
+    
+    public init(allowLiterals: Bool? = nil,
+                 namedBindings: GqlQueryNamedBindings? = nil,
+                 positionalBindings: [GqlQueryParameter]? = nil,
+                 queryString: String? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Query.swift [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        switch gqlQueryParameter {
+        case .value(let value):
+            self.init(value: value)
+        case .cursor(let string):
+            self.init(cursor: string)
+        }
+    }
+    
+    init(cursor: String? = nil,
+                 value: Value? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Query.swift]: indentation

File: Datastore/Sources/Data API/Models/Project API/Types/Value.swift [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                 doubleValue: Double? = nil,
+                 entityValue: Entity? = nil,
+                 geoPointValue: LatLng? = nil,
+                 integerValue: String? = nil,
+                 keyValue: Key? = nil,
+                 nullValue: String? = nil,
+                 stringValue: String? = nil,
+                 timestampValue: String? = nil,
+                 excludeFromIndexes: Bool? = nil,
+                 meaning: Int? = nil) {
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/Models/Project API/Types/Value.swift]: Indentation

File: Datastore/Sources/DatastoreClient.swift [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import Foundation
+import AsyncHTTPClient
+import NIO
+
+public final class GoogleCloudDatastoreClient {
+
+    public var project: DatastoreProjectAPI
+    var datastoreRequest: GoogleCloudDatastoreRequest
+    
+    /// Initialize a client for interacting with the Google Cloud Storage API
```
Comment (Andrewangeta) [inline Datastore/Sources/DatastoreClient.swift]: Just some cloud storage verbiage in here.

File: Datastore/Sources/DatastoreClient.swift [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        guard let projectId = ProcessInfo.processInfo.environment["PROJECT_ID"] ??
+                                (refreshableToken as? OAuthServiceAccount)?.credentials.projectId ??
+                                config.project ?? credentials.project else {
+            throw GoogleCloudDatastoreError.projectIdMissing
+        }
+
+        datastoreRequest = GoogleCloudDatastoreRequest(httpClient: httpClient,
+                                                   eventLoop: eventLoop,
+                                                   oauth: refreshableToken,
+                                                   project: projectId)
```
Comment (Andrewangeta) [inline Datastore/Sources/DatastoreClient.swift]: Indentation

File: Datastore/Sources/DatastoreConfig.swift [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        self.scope = scope
+        self.serviceAccount = serviceAccount
+        self.project = project
+    }
+    
+    /// Create a new `GoogleCloudDatastoreConfiguration` with datastore scope and the default service account.
+    public static func `default`() -> GoogleCloudDatastoreConfiguration {
+        return GoogleCloudDatastoreConfiguration(scope: [.datastore],
+                                               serviceAccount: "default",
+                                               project: nil)
```
Comment (Andrewangeta) [inline Datastore/Sources/DatastoreConfig.swift]: Indentation

File: Datastore/Sources/DatastoreConfig.swift [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+public enum GoogleCloudDatastoreScope: GoogleCloudAPIScope {
+    /// View and manage your Google Cloud Datastore data
+    case datastore
+    /// View and manage your data across Google Cloud Platform services
+    case cloudPlatform
+    
+    public var value: String {
+        switch self {
+            case .datastore: return "https://www.googleapis.com/auth/datastore"
+            case .cloudPlatform: return "https://www.googleapis.com/auth/cloud-platform"
```
Comment (Andrewangeta) [inline Datastore/Sources/DatastoreConfig.swift]: Indentation

**Review by Andrewangeta — CHANGES_REQUESTED** [review]
This mostly looks good to me! Thanks for the PR :)
Other than the comments about indentation.

File: Datastore/Sources/Data API/API/ProjectAPI.swift [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    
+    public func allocateIDs(keys: [Key]) -> EventLoopFuture<AllocateIdsResponse> {
+        
+        do {
+            let allocateIdsRequest = AllocateIdsRequest(keys: keys)
+            let body = try HTTPClient.Body.data(encoder.encode(allocateIdsRequest))
+            return request.send(method: .POST, path: "\(projectPath):allocateIds", body: body)
+        } catch {
+            return request.httpClient.eventLoopGroup.next().makeFailedFuture(error)
```
Comment (Andrewangeta) [inline Datastore/Sources/Data API/API/ProjectAPI.swift]: Also we can actually just use `request.eventLoop` for all the catch blocks in the requests and not have to ask the client anymore. Some of this in the cloud storage API was leftover and I'll raise a PR to clean it up soon.

**Review by Andrewangeta — CHANGES_REQUESTED** [review]

**Comment by owainhunt** [comment]
All fixed! Corresponding change for vapor-community/google-cloud ready to go once this is merged in.

File: Datastore/Sources/DatastoreClient.swift [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        /// A token implementing `OAuthRefreshable`. Loaded from credentials specified by `GoogleCloudCredentialsConfiguration`.
+        let refreshableToken = OAuthCredentialLoader.getRefreshableToken(credentials: credentials,
+                                                                         withConfig: config,
+                                                                         andClient: httpClient,
+                                                                         eventLoop: eventLoop)
+
+        /// Set the projectId to use for this client. In order of priority:
+        /// - Environment Variable (PROJECT_ID)
+        /// - Service Account's projectID
+        /// - `GoogleCloudStorageConfigurations` `project` property (optionally configured).
```
Comment (Andrewangeta) [inline Datastore/Sources/DatastoreClient.swift]: Should be `GoogleCloudDatastoreConfiguration`.

File: Datastore/Sources/DatastoreClient.swift (hunk 11 above)
Comment (Andrewangeta) [reply Datastore/Sources/DatastoreClient.swift]: Small indentation here.

**Review by Andrewangeta — CHANGES_REQUESTED** [review]
Just a couple of indentations.

**Comment by owainhunt** [comment]
Fixed.

**Comment by Andrewangeta** [comment]
@Owainhunt thanks! And as a final sanity check, did you run locally to test the API works as expected?

**Comment by owainhunt** [comment]
Just added something I didn't stage in the last commit. All good now - confirmed I've tested locally and all works as expected.

**Comment by Andrewangeta** [comment]
Awesome!!

**Review by Andrewangeta — APPROVED** [review]

**Pull request merged**